tests/unit/utils/test_nova_compute.py

```diff
@@ -0,0 +1,87 @@
+#  Copyright 2023 Canonical Limited
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+from unittest.mock import AsyncMock, MagicMock, patch
+
+import pytest
+from juju.action import Action
+
+from cou.apps.base import ApplicationUnit
+from cou.utils import nova_compute
+from tests.unit.conftest import _generate_mock_machine
+
+
+@pytest.mark.asyncio
+async def test_get_instance_count(model):
+    expected_count = 1
+    model.run_action.return_value = mocked_action = AsyncMock(spec_set=Action).return_value
+    mocked_action.results = {"Code": "0", "instance-count": str(expected_count)}
+
+    actual_count = await nova_compute.get_instance_count(unit="nova-compute/0", model=model)
+
+    model.run_action.assert_called_once_with(
+        unit_name="nova-compute/0",
+        action_name="instance-count",
+    )
+    assert actual_count == expected_count
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "result_key, value",
+    [
+        ("not_valid", "1"),  # invalid key
+        ("instance-count", "not_valid"),  # invalid value
+        ("not_valid", "not_valid"),  # invalid key and value
+    ],
+)
+async def test_get_instance_count_invalid_result(model, result_key, value):
+    model.run_action.return_value = mocked_action = AsyncMock(spec_set=Action).return_value
+    mocked_action.results = {"Code": "0", result_key: value}
+
+    with pytest.raises(ValueError):
+        await nova_compute.get_instance_count(unit="nova-compute/0", model=model)
+
+
+@pytest.mark.parametrize(
+    "hypervisors_count, expected_result",
+    [
+        ([(0, 0), (1, 0), (2, 0)], {"0", "1", "2"}),
+        ([(0, 1), (1, 0), (2, 0)], {"1", "2"}),
+        ([(0, 1), (1, 3), (2, 0)], {"2"}),
+        ([(0, 1), (1, 3), (2, 5)], set()),
+    ],
+)
+@pytest.mark.asyncio
+@patch("cou.utils.nova_compute.get_instance_count")
+@patch("cou.utils.nova_compute.asyncio.gather", new_callable=AsyncMock)
```

Comment:
not-blocking: IMO, it would be cleaner to define `mock_instance_count.side_effect` and do not mock the gather here.